### ISSUE TITLE
Training on A2D2 Dataset

### ISSUE THREAD

**Issue by shu7725** [nianticlabs/monodepth2#148]
Hello,

First of all, thanks for the code.

I am trying to train the network with Audi dataset. I am using the images from the front-center configuration. Original image resolution is 1920x1208. Image extension is .png.
Here are the steps that I followed,
1. Created train_files.txt and val_files.txt by selecting every middle frame.(1,3,5 ....)
2. Remove the sequence with very little motion between the frames from train_files and val_files.
3. Resize the images to (1024x320) and (640x192)
4. Created a training class which inherits class MonoDataset.
5. ResNet18, Batch Size 4, Learning rate 1e-5, Number of epochs 20. 
6. self.K = np.array([[focal / image_width, 0, 0.5, 0],
                           [0, focal / image_height, 0.5, 0],
                           [0, 0, 1, 0],
                           [0, 0, 0, 1]], dtype=np.float32)
where, focal = image_width / (2 * np.tan(FOV * np.pi / 360))
            image_width=1024(Resized width) and FOV=60(FOV from dataset)

Here are the logs from my training,

Training model named:
   a2d2_1024x320
Models and tensorboard events files are saved to:
   C:\Users\aimachine\tmp
Training is using:
   cuda
Using split:
   A2D2

There are 10012 training items and 3568 validation items

Training

epoch   0 | batch      0 | examples/s:   0.9 | loss: 0.12477 | time elapsed: 00h00m20s | time left: 00h00m00s
epoch   0 | batch    250 | examples/s:   4.9 | loss: 0.10537 | time elapsed: 00h03m56s | time left: 13h05m28s
epoch   0 | batch    500 | examples/s:   5.0 | loss: 0.08677 | time elapsed: 00h07m33s | time left: 12h28m22s
epoch   0 | batch    750 | examples/s:   4.9 | loss: 0.08290 | time elapsed: 00h11m09s | time left: 12h13m41s
epoch   0 | batch   1000 | examples/s:   4.8 | loss: 0.09518 | time elapsed: 00h14m45s | time left: 12h04m04s
epoch   0 | batch   1250 | examples/s:   4.9 | loss: 0.09022 | time elapsed: 00h18m21s | time left: 11h56m46s
epoch   0 | batch   1500 | examples/s:   4.8 | loss: 0.08835 | time elapsed: 00h21m57s | time left: 11h50m52s
epoch   0 | batch   1750 | examples/s:   4.9 | loss: 0.07992 | time elapsed: 00h25m33s | time left: 11h45m32s
epoch   0 | batch   2000 | examples/s:   4.9 | loss: 0.10009 | time elapsed: 00h29m09s | time left: 11h40m36s
Training
epoch   1 | batch   1497 | examples/s:   5.0 | loss: 0.08082 | time elapsed: 00h56m45s | time left: 10h53m30s
Training
epoch   2 | batch    994 | examples/s:   4.9 | loss: 0.09052 | time elapsed: 01h24m20s | time left: 10h19m23s
Training
epoch   3 | batch    491 | examples/s:   5.0 | loss: 0.05996 | time elapsed: 01h51m59s | time left: 09h48m49s
epoch   3 | batch   2491 | examples/s:   4.9 | loss: 0.06339 | time elapsed: 02h19m20s | time left: 09h18m12s
Training
epoch   4 | batch   1988 | examples/s:   5.0 | loss: 0.06198 | time elapsed: 02h46m56s | time left: 08h49m28s
Training
epoch   5 | batch   1485 | examples/s:   4.8 | loss: 0.06003 | time elapsed: 03h14m32s | time left: 08h21m06s
Training
epoch   6 | batch    982 | examples/s:   5.0 | loss: 0.06871 | time elapsed: 03h42m09s | time left: 07h52m55s
Training
epoch   7 | batch    479 | examples/s:   4.9 | loss: 0.06467 | time elapsed: 04h09m48s | time left: 07h24m56s
epoch   7 | batch   2479 | examples/s:   4.8 | loss: 0.06328 | time elapsed: 04h37m09s | time left: 06h56m34s
Training
epoch   8 | batch   1976 | examples/s:   4.9 | loss: 0.06840 | time elapsed: 05h04m46s | time left: 06h28m42s
Training
epoch   9 | batch   1473 | examples/s:   4.9 | loss: 0.06086 | time elapsed: 05h32m22s | time left: 06h00m53s
Training
epoch  10 | batch    970 | examples/s:   4.9 | loss: 0.08983 | time elapsed: 05h59m57s | time left: 05h33m05s
Training
epoch  11 | batch    467 | examples/s:   4.9 | loss: 0.06303 | time elapsed: 06h27m30s | time left: 05h05m18s
epoch  11 | batch   2467 | examples/s:   4.8 | loss: 0.06010 | time elapsed: 06h54m50s | time left: 04h37m23s
Training
epoch  12 | batch   1964 | examples/s:   4.7 | loss: 0.06366 | time elapsed: 07h22m26s | time left: 04h09m42s
Training
epoch  13 | batch   1461 | examples/s:   4.9 | loss: 0.06610 | time elapsed: 07h50m02s | time left: 03h42m01s
Training
epoch  14 | batch    958 | examples/s:   4.9 | loss: 0.06007 | time elapsed: 08h17m38s | time left: 03h14m21s
Training
epoch  15 | batch    455 | examples/s:   4.9 | loss: 0.06917 | time elapsed: 08h45m14s | time left: 02h46m41s
epoch  15 | batch   2455 | examples/s:   4.9 | loss: 0.07661 | time elapsed: 09h12m35s | time left: 02h18m58s
Training
epoch  16 | batch   1952 | examples/s:   4.9 | loss: 0.05360 | time elapsed: 09h40m12s | time left: 01h51m20s
Training
epoch  17 | batch   1449 | examples/s:   4.9 | loss: 0.07085 | time elapsed: 10h07m47s | time left: 01h23m42s
Training
epoch  18 | batch    946 | examples/s:   4.9 | loss: 0.05948 | time elapsed: 10h35m23s | time left: 00h56m04s
Training
epoch  19 | batch    443 | examples/s:   4.9 | loss: 0.06000 | time elapsed: 11h03m00s | time left: 00h28m27s
epoch  19 | batch   2443 | examples/s:   4.9 | loss: 0.07037 | time elapsed: 11h30m20s | time left: 00h00m49s

Process finished with exit code 0

As you see, the loss is consistant and not reducing. Am I doing something wrong ? Can you please help me ?

(GPU- NVIDIA GTX 2080 - TI)

**Comment by mdfirman** [comment]
Hi, thanks for your interest! This question is a little outside our standard scope, but I can provide some basic advice.

The loss you have reported looks ok, sometimes it goes up and down a bit.

The most important thing is: what does your tensorboard look like? Do you have plausible looking depth maps?

**Comment by shu7725** [comment]
Hello,

Thanks for the reply.

The output for one of the frames looks like this,

![Output](https://user-images.githubusercontent.com/57450116/75558796-a8874480-5a42-11ea-94d0-05f9ed95e21f.PNG)

I tried increasing the epochs to 50. Results did not change much.
I am using opencv default resize option to resize the input images to defined resolution(640x192). 

Would you suggest any changes to input..? 

Thanks.

**Comment by mrharicot** [comment]
Hi,
It is difficult to assess what is going on without really looking at the data.
Here are some suggestions:
- How much time and motion is there between consecutive frames?
- 10k training frames is fairly small
- 1024x320 is very large to run some initial experiments, I would suggest to reduce it to something smaller and keep the aspect ratio close to the original data like 512x320. This will significantly speed up training and will allow you to increase the batch size, improving stability. 
- To downsample you should either use INTER_AREA in opencv or imagemagick which I believe does a good job at dowsampling

**Comment by caikw0602** [comment]
Hi, shu7725. I wanna train my own monocular dataset with monodepth2 too. But I am struggling how to make the right data orgarization. Further, I do not know how to name the images of different frames. Would you please give some help? Some work steps? Thank you very much.

**Comment by shu7725** [comment]
Hello @caikw0602 

I am not sure about the type of data that your have. 
So far I have done following modification to train my dataset,
1. Acquire dataset and extract file names into train.txt and val.txt according to your split.
2. Reshape the images suggested resolutions.
3. Create a class for your data which inherits the class "MonoDataset"
4. Assuming you dont have GT Depth values, set get_depth function always returns false.
5. Modify the value of K according to your data.
6. train.txt and val.txt contains every middle frame of dataset. (for ex, assuming you have files in sequence 0,1,2,3.. your train.txt would contain 1,3, and so on)


@mrharicot Can you please verify the steps..? It might help us both..

Thanks in advance.:)

**Comment by mrharicot** [comment]
@shu7725 @caikw0602 looks good to me!

**Comment by caikw0602** [comment]
Thank you very much. I would try with your reply.

**Comment by floydium** [comment]
Hi @shu7725,

I think you should not split the dataset in this alternating fashion. A better way would be to sort them by time (or frame number) and assign the first portion to the test set and the second portion to the validation set. This way you are not spilling training information into your validation set. It is not like you are overfitting, but I would keep an eye on this. 

AFAIK the A2D2 dataset has about 390k frames - try throwing more more data into it, and see if that helps.

**Comment by shu7725** [comment]
Hello @floydium,
Thanks for the comment. I will try again with your suggestions.
:)

**Comment by Jinho-P** [comment]
Hi @shu7725,

I wonder if you have any advanced result of above problem.
Could you answer to me about any advice?
